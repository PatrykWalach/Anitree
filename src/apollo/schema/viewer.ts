export interface User {
  id: number
  name: string
  avata: UserAvatar
  siteUrl: string
  options: {
    profileColor: string
    displayAdultContent: boolean
  }
  mediaListOptions: {
    scoreFormat: ScoreFormat
    animeList: MediaListTypeOptions
    mangaList: MediaListTypeOptions
  }
}
export interface UserAvatar {
  large: string
  medium: string
}
export type ScoreFormat =
  | 'POINT_100'
  | 'POINT_10_DECIMAL'
  | 'POINT_10'
  | 'POINT_5'
  | 'POINT_3'

export interface MediaListTypeOptions {
  customLists: string[]
  advancedScoring: string[]
  advancedScoringEnabled: boolean
}
