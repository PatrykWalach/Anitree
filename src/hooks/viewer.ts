import { inject } from '@vue/composition-api'
import { useQuery } from '@vue/apollo-composable'
export const DefaultViewer = Symbol('Viewer')

export const useViewer = <V = ReturnType<typeof useQuery>>() =>
  inject(DefaultViewer) as V
