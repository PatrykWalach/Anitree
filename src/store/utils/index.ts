export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
