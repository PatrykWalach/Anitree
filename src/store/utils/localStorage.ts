export const getItem = <T>(key: string, defaultValue?: T) => {
  const cache = localStorage.getItem(key)
  const stored = cache && JSON.parse(cache)

  return {
    ...defaultValue,
    ...stored,
  } as T
}
export const setItem = <T>(key: string, value: T) =>
  localStorage.setItem(key, JSON.stringify({ ...value }))
