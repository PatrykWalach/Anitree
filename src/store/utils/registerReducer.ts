// import * as reducers from '../reducers'
// import { Reducer, ReducersMapObject, combineReducers } from 'redux'
// import { EnhancedStore } from '@reduxjs/toolkit'
// import { store, useStore } from '../'

// const staticReducers: ReducersMapObject<any, any> = reducers
// const asyncReducers: ReducersMapObject<any, any> = {}

// export const registerReducer = <K extends string, R extends Reducer<any>>(
//   key: K,
//   asyncReducer: R,
// ) => {
//   asyncReducers[key] = asyncReducer
//   store.replaceReducer(createReducer(asyncReducers) as any)

//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<S & { [key in K]: R }>
//         : any
//     >()
// }

// export const unregisterReducer = <K extends string>(key: K) => {
//   delete asyncReducers[key]
//   store.replaceReducer(createReducer(asyncReducers) as any)
//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<Omit<S, K>>
//         : any
//     >()
// }

// const createReducer = (asyncReducers: ReducersMapObject<any, any> = {}) =>
//   combineReducers({
//     ...staticReducers,
//     ...asyncReducers,
//   })
