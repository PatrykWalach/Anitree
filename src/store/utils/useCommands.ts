import { ListCommand } from '../modules/commands'
import { computed } from '@vue/composition-api'
import { useState } from './useState'
import { useStore } from './useStore'

export const useCommands = () => {
  const { dispatch, actions } = useStore()

  const history = useState(state => state.commands.history)

  const undo = async () => {
    history.value[history.value.length - 1]
      .undo()
      .then(() => dispatch(actions.commands.POP_COMMAND()))
  }

  const add = async (command: ListCommand) => {
    dispatch(actions.commands.PUSH_COMMAND(command))
    await command.execute()
    return command
  }

  const getPending = computed(() => history.value.filter(({ done }) => !done))
  return { add, getPending, undo }
}
