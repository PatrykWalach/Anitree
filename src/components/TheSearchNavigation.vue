<template>
  <v-container>
    <v-row justify="center" align="start">
      <v-col cols="12" md="6">
        <v-card>
          <v-list subheader>
            <v-subheader>
              Here are just some of the things you can see
            </v-subheader>
            <v-list-item
              v-for="{ bind, title, icon } of search"
              :key="title"
              v-bind="bind"
              replace
            >
              <v-list-item-icon>
                <v-icon>{{ icon }}</v-icon>
              </v-list-item-icon>
              <v-list-item-content>
                <v-list-item-title>{{ title }}</v-list-item-title>
              </v-list-item-content>
            </v-list-item>
          </v-list>
        </v-card>
      </v-col>
    </v-row></v-container
  >
</template>

<script lang="ts">
import { NavigationElement } from '../types'
import { User } from '@/graphql/schema/viewer'
import { createComponent } from '@vue/composition-api'

export interface Props {
  user: null | User
}
export const useNavigation = () => {
  const search: NavigationElement[] = [
    {
      bind: {
        to: {
          name: 'search',
          query: {
            season: (() => {
              const months = ['WINTER', 'SPRING', 'SUMMER', 'FALL']
              const date = Math.floor(new Date().getMonth() / 3)
              return months[date]
            })(),
            sort: 'POPULARITY_DESC',
            year: new Date().getFullYear().toString(),
          },
        },
      },
      icon: 'whatshot',
      title: 'Current season',
    },
    {
      bind: {
        to: {
          name: 'search',
          query: {
            sort: 'ID_DESC',
          },
        },
      },
      icon: 'new_releases',
      title: 'Recently added',
    },
    {
      bind: {
        to: {
          name: 'search',
          query: {
            sort: 'TRENDING_DESC',
          },
        },
      },
      icon: 'trending_up',
      title: 'Trending',
    },
  ]
  return { search }
}

export default createComponent<Readonly<Props>>({
  props: {
    user: { default: null, required: true, type: null },
  },
  setup() {
    return useNavigation()
  },
})
</script>
