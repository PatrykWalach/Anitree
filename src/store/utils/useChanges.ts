import {
  DELETE_MEDIA_LIST_ENTRY,
  MEDIA,
  SAVE_MEDIA_LIST_ENTRY,
} from '@/graphql'
import { DeletePending, SavePending } from '../modules/changes'
import {
  DeleteVariables,
  Deleted,
  SaveVariables,
} from '@/graphql/schema/listEntry'
import { DocumentNode, FetchResult } from 'apollo-link'
import { Ref, ref } from '@vue/composition-api'
import { DataProxy } from 'apollo-cache'
import { Media } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import { MutationUpdaterFn } from 'apollo-client'
import { Variables } from '@/graphql/schema/media'
import { useActions } from '..'
import { useDispatch } from 'vue-redux-hooks'
import { useMutation } from '@vue/apollo-composable'

type OnError = (
  fn: (param?: Error | undefined) => void,
) => {
  off: () => void
}

export const handleError = <
  P extends Omit<SavePending, 'id'> | Omit<DeletePending, 'id'>
>(
  onError: OnError,
  type: P['type'],
) => {
  const pending: Ref<P> = ref({ type })

  const dispatch = useDispatch()
  const actions = useActions()

  onError(() => {
    dispatch(actions.changes.UNSHIFT_PENDING(pending.value))
  })

  return pending
}

export const useChange = <D, V>(
  query: DocumentNode,
  update: (cache: DataProxy, { data }: FetchResult<D>, mediaId: number) => void,
) => {
  const { mutate: change, loading, onError, onDone } = useMutation<D, V>(query)
  const mutate = (variables: V, mediaId: number) => {
    change(variables, {
      update: (cache, result) => update(cache, result, mediaId),
    })
  }
  return { loading, mutate, onDone, onError }
}

export const useDelete = () => {
  return useChange<{ DeleteMediaListEntry: Deleted }, DeleteVariables>(
    DELETE_MEDIA_LIST_ENTRY,
    updateDeleteMediaListEntry,
  )
}

export const useSave = () => {
  return useChange<
    {
      SaveMediaListEntry: MediaList
    },
    SaveVariables
  >(SAVE_MEDIA_LIST_ENTRY, updateSaveMediaListEntry)
}

export const useDeleteWithChanges = () => {
  const query = useDelete()

  const pending = handleError(query.onError, 'DELETE')

  const mutate = (variables: DeleteVariables, mediaId: number) => {
    pending.value = {
      mediaId,
      type: 'DELETE',
      variables,
    }

    query.mutate(variables, mediaId)
  }

  return { ...query, mutate }
}
export const useSaveWithChanges = () => {
  const query = useSave()

  const pending = handleError(query.onError, 'SAVE')

  const mutate = (variables: SaveVariables) => {
    pending.value = {
      mediaId: variables.mediaId,
      type: 'SAVE',
      variables,
    }

    query.mutate(variables, variables.mediaId)
  }

  return { ...query, mutate }
}

export const useChanges = () => {
  const { mutate: deleteEntry } = useDelete()
  const { mutate: saveEntry } = useSave()

  const isSavePending = (
    pending: SavePending | DeletePending,
  ): pending is SavePending => pending.type === 'SAVE'

  return { deleteEntry, isSavePending, saveEntry }
}

export const updateSaveMediaListEntry: MutationUpdaterFn<{
  SaveMediaListEntry: MediaList
}> = (cache, { data: result }) => {
  if (!result) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: { id: result.SaveMediaListEntry.mediaId },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = result.SaveMediaListEntry

    cache.writeQuery({
      ...query,
      data,
    })
  }
}

const updateDeleteMediaListEntry = (
  cache: DataProxy,
  { data: result }: FetchResult<{ DeleteMediaListEntry: Deleted }>,
  id: number,
) => {
  if (!result || !result.DeleteMediaListEntry.deleted) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: {
      id,
    },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = null

    cache.writeQuery({
      ...query,
      data,
    })
  }
}
