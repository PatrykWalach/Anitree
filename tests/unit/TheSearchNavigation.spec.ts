// user: null | User

import { createVariations, matchSnapshot } from './utils'
import TheSearchNavigation from '@/components/TheSearchNavigation.vue'
import { mockedViewer } from './mocks/viewer'

describe('TheSearchNavigation', () => {
  createVariations({
    propsData: {
      user: () => [null, mockedViewer],
    },
  }).forEach(settings => matchSnapshot(TheSearchNavigation, settings))
})
