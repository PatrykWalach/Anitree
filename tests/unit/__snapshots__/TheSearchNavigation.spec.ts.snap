// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheSearchNavigation matches snapshot 1`] = `
<v-container-stub
  tag="div"
>
  <v-row-stub
    align="start"
    justify="center"
    tag="div"
  >
    <v-col-stub
      cols="12"
      md="6"
      tag="div"
    >
      <v-card-stub
        loaderheight="4"
        tag="div"
      >
        <v-list-stub
          subheader="true"
          tag="div"
          tile="true"
        >
          <v-subheader-stub>
            
            Here are just some of the things you can see
          
          </v-subheader-stub>
           
          <v-list-item-stub
            activeclass=""
            replace="true"
            tag="div"
            to="[object Object]"
          >
            <v-list-item-icon-stub>
              <v-icon-stub>
                whatshot
              </v-icon-stub>
            </v-list-item-icon-stub>
             
            <v-list-item-content-stub>
              <v-list-item-title-stub>
                Current season
              </v-list-item-title-stub>
            </v-list-item-content-stub>
          </v-list-item-stub>
          <v-list-item-stub
            activeclass=""
            replace="true"
            tag="div"
            to="[object Object]"
          >
            <v-list-item-icon-stub>
              <v-icon-stub>
                new_releases
              </v-icon-stub>
            </v-list-item-icon-stub>
             
            <v-list-item-content-stub>
              <v-list-item-title-stub>
                Recently added
              </v-list-item-title-stub>
            </v-list-item-content-stub>
          </v-list-item-stub>
          <v-list-item-stub
            activeclass=""
            replace="true"
            tag="div"
            to="[object Object]"
          >
            <v-list-item-icon-stub>
              <v-icon-stub>
                trending_up
              </v-icon-stub>
            </v-list-item-icon-stub>
             
            <v-list-item-content-stub>
              <v-list-item-title-stub>
                Trending
              </v-list-item-title-stub>
            </v-list-item-content-stub>
          </v-list-item-stub>
        </v-list-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;

exports[`TheSearchNavigation matches snapshot 2`] = `
<v-container-stub
  tag="div"
>
  <v-row-stub
    align="start"
    justify="center"
    tag="div"
  >
    <v-col-stub
      cols="12"
      md="6"
      tag="div"
    >
      <v-card-stub
        loaderheight="4"
        tag="div"
      >
        <v-list-stub
          subheader="true"
          tag="div"
          tile="true"
        >
          <v-subheader-stub>
            
            Here are just some of the things you can see
          
          </v-subheader-stub>
           
          <v-list-item-stub
            activeclass=""
            replace="true"
            tag="div"
            to="[object Object]"
          >
            <v-list-item-icon-stub>
              <v-icon-stub>
                whatshot
              </v-icon-stub>
            </v-list-item-icon-stub>
             
            <v-list-item-content-stub>
              <v-list-item-title-stub>
                Current season
              </v-list-item-title-stub>
            </v-list-item-content-stub>
          </v-list-item-stub>
          <v-list-item-stub
            activeclass=""
            replace="true"
            tag="div"
            to="[object Object]"
          >
            <v-list-item-icon-stub>
              <v-icon-stub>
                new_releases
              </v-icon-stub>
            </v-list-item-icon-stub>
             
            <v-list-item-content-stub>
              <v-list-item-title-stub>
                Recently added
              </v-list-item-title-stub>
            </v-list-item-content-stub>
          </v-list-item-stub>
          <v-list-item-stub
            activeclass=""
            replace="true"
            tag="div"
            to="[object Object]"
          >
            <v-list-item-icon-stub>
              <v-icon-stub>
                trending_up
              </v-icon-stub>
            </v-list-item-icon-stub>
             
            <v-list-item-content-stub>
              <v-list-item-title-stub>
                Trending
              </v-list-item-title-stub>
            </v-list-item-content-stub>
          </v-list-item-stub>
        </v-list-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
